internal/pprof/pprof.go

```diff
@@ -0,0 +1,59 @@
+package pprof
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime/pprof"
+)
+
+type profileSession struct {
+	cpuFilePath string
+	memFilePath string
+	cpuFile     *os.File
+	memFile     *os.File
+}
+
+func BeginProfiling(profileDir string) *profileSession {
+	if err := os.MkdirAll(profileDir, 0o755); err != nil {
+		panic(err)
+	}
+
+	pid := os.Getpid()
+
+	cpuProfilePath := filepath.Join(profileDir, fmt.Sprintf("%d-cpuprofile.pb.gz", pid))
+	memProfilePath := filepath.Join(profileDir, fmt.Sprintf("%d-memprofile.pb.gz", pid))
+	cpuFile, err := os.Create(cpuProfilePath)
+	if err != nil {
+		panic(err)
+	}
+	memFile, err := os.Create(memProfilePath)
+	if err != nil {
+		panic(err)
+	}
+
+	if err := pprof.StartCPUProfile(cpuFile); err != nil {
+		panic(err)
+	}
+
+	return &profileSession{
+		cpuFilePath: cpuProfilePath,
+		memFilePath: memProfilePath,
+		cpuFile:     cpuFile,
+		memFile:     memFile,
+	}
+}
+
+func (p *profileSession) Stop() {
+	pprof.StopCPUProfile()
+	err := pprof.Lookup("allocs").WriteTo(p.memFile, 0)
+	if err != nil {
+		panic(err)
+	}
+
+	p.cpuFile.Close()
+	p.memFile.Close()
+
+	fmt.Printf("CPU profile: %v\n", p.cpuFilePath)
+	fmt.Printf("Memory profile: %v\n", p.memFilePath)
```

Comment:
These should change to `fmt.Fprintf(os.Stderr, ...)`, otherwise the LS will break due to us writing at the end to the JSON-RPC pipe.